flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1429,61 +1390,45 @@ public void run() {
 		}
 	}
 
-	private void failPendingCheckpoint(
-			final PendingCheckpoint pendingCheckpoint,
-			final CheckpointFailureReason reason) {
-
-		failPendingCheckpoint(pendingCheckpoint, reason, null);
-	}
-
-	private void failPendingCheckpoint(
-		final PendingCheckpoint pendingCheckpoint,
-		final CheckpointFailureReason reason,
-		@Nullable final Throwable cause) {
-
-		CheckpointException exception = new CheckpointException(reason, cause);
-		pendingCheckpoint.abort(reason, cause);
-		failureManager.handleJobLevelCheckpointException(exception, pendingCheckpoint.getCheckpointId());
-
-		checkAndResetCheckpointScheduler();
-	}
-
-	private void failPendingCheckpointDueToTaskFailure(
-		final PendingCheckpoint pendingCheckpoint,
-		final CheckpointFailureReason reason,
-		final ExecutionAttemptID executionAttemptID) {
+	private void abortPendingCheckpoint(
+		PendingCheckpoint pendingCheckpoint,
+		CheckpointException exception) {
 
-		failPendingCheckpointDueToTaskFailure(pendingCheckpoint, reason, null, executionAttemptID);
+		abortPendingCheckpoint(pendingCheckpoint, exception, null);
 	}
 
-	private void failPendingCheckpointDueToTaskFailure(
-			final PendingCheckpoint pendingCheckpoint,
-			final CheckpointFailureReason reason,
-			@Nullable final Throwable cause,
-			final ExecutionAttemptID executionAttemptID) {
-
-		CheckpointException exception = new CheckpointException(reason, cause);
-		pendingCheckpoint.abort(reason, cause);
-		failureManager.handleTaskLevelCheckpointException(exception, pendingCheckpoint.getCheckpointId(), executionAttemptID);
+	private void abortPendingCheckpoint(
+		PendingCheckpoint pendingCheckpoint,
+		CheckpointException exception,
+		@Nullable final ExecutionAttemptID executionAttemptID) {
 
-		checkAndResetCheckpointScheduler();
-	}
+		assert(Thread.holdsLock(lock));
 
-	private void checkAndResetCheckpointScheduler() {
-		if (!shutdown && periodicScheduling && currentPeriodicTrigger == null) {
-			synchronized (lock) {
-				if (pendingCheckpoints.isEmpty() || allPendingCheckpointsDiscarded()) {
-					triggerRequestQueued = false;
-					currentPeriodicTrigger = scheduleTriggerWithDelay(getRandomInitDelay());
```

Comment:
what has happened to this logic?

Reply:
Basically `checkAndResetCheckpointScheduler` does same thing with `triggerQueuedRequests` (try to resume the periodic triggering) but under a stricter condition (there is no more in-flight checkpoint). 
`checkAndResetCheckpointScheduler` is invoked when an in-flight checkpoint fails. `triggerRequestQueued` is invoked when an in-flight checkpoint succeeds or timeouts.

It does not makes much sense but it makes the codes more complicated. I think the rule here could be simplified a lot.

It suspends the periodic triggering when there are too many in-flight checkpoints. So it should resume periodic triggering when there are fewer in-flight checkpoints than `maxConcurrentCheckpointAttempts`. That means trying to resume periodic triggering when an in-flight checkpoint ends its lifecycle, no matter succeeding or failing.

Reply:
ok, it makes sense to me.